components/tidb_query/src/codec/mysql/json/comparison.rs

```diff
@@ -19,25 +19,26 @@ const PRECEDENCE_STRING: i32 = -10;
 const PRECEDENCE_NUMBER: i32 = -11;
 const PRECEDENCE_NULL: i32 = -12;
 
-impl Json {
+impl<'a> JsonRef<'a> {
     fn get_precedence(&self) -> i32 {
-        match *self {
-            Json::Object(_) => PRECEDENCE_OBJECT,
-            Json::Array(_) => PRECEDENCE_ARRAY,
-            Json::Boolean(_) => PRECEDENCE_BOOLEAN,
-            Json::None => PRECEDENCE_NULL,
-            Json::I64(_) | Json::U64(_) | Json::Double(_) => PRECEDENCE_NUMBER,
-            Json::String(_) => PRECEDENCE_STRING,
+        match self.get_type() {
+            JsonType::Object => PRECEDENCE_OBJECT,
+            JsonType::Array => PRECEDENCE_ARRAY,
+            JsonType::Literal => self
+                .get_literal()
+                .map_or(PRECEDENCE_NULL, |_| PRECEDENCE_BOOLEAN),
+            JsonType::I64 | JsonType::U64 | JsonType::Double => PRECEDENCE_NUMBER,
+            JsonType::String => PRECEDENCE_STRING,
         }
     }
 
     fn as_f64(&self) -> Result<f64> {
-        match *self {
-            Json::Double(d) => Ok(d),
-            Json::I64(d) => Ok(d as f64),
-            Json::U64(d) => Ok(d as f64),
-            Json::Boolean(_) => {
-                let v = self.as_literal()?;
+        match self.get_type() {
+            JsonType::I64 => Ok(self.get_i64() as f64),
+            JsonType::U64 => Ok(self.get_u64() as f64),
+            JsonType::Double => Ok(self.get_double()),
+            JsonType::Literal => {
+                let v = self.as_literal().unwrap();
```

Comment:
Is it safe to unwrap here?

Reply:
Since it's a `JsonType::Literal`, the `unwrap()` here is safe